### DISCUSSION TITLE
More buttons 64 instead of 32

### DISCUSSION THREAD

**Discussion by Pawan419** [vsulako/AFFBWheel#40] [Show and tell]
Hey. Can anyone help me for increasing Buttons.
I need 64 buttons using 8 shift registers. M not good in coding so can anyone help me in it

**Comment by vsulako** [comment]
I thought 32 buttons should be enough for any racing simulator and chose 32 because of fewer ICs needed and slightly faster polling. :)

What is needed to change to  64 buttons...
Sorry, I cannot test it in hardware right now, so the following is pure theoretical:

1) change HID descriptor to report 64 buttons instead of 32:

hidDescriptor.h, lines 32-42:

```
      //64 buttons
      0x05, 0x09,             // USAGE_PAGE (Button)
      0x19, 0x01,             // Usage Minimum (1),
      0x29, 0x40,             // Usage Maximum (64)  - change here
      0x15, 0x00,             // LOGICAL_MINIMUM (0)
      0x25, 0x01,             // Logical Maximum (1),
      0x35, 0x00,             // PHYSICAL_MINIMUM (0) 
      0x45, 0x01,             // Physical Maximum (1),
      0x95, 0x40,             // Report Count (64 fields), - and here
      0x75, 0x01,             // Report Size (1 bit),
      0x81, 0x02,             // INPUT (Data,Var,Abs) 
```

2) change HID report structure to match HID descriptor having 64bits for buttons now:

wheel.h, line 20:

`  uint64_t buttons;`
  
3) also change temporary variables for buttons:

wheel.h, line 90:

`  uint64_t buttons;`

AFFBWheel.ino, line 61:

` uint64_t tempButtons;`

line 804:

` uint64_t buttons=0;`

4) I assume you use 74HC165 shift registers, in 2 channels of 4 chained registers each (instead of 2 registers per channel). 
Change `readButtons` procedure to read 2x more data:

in AFFBWheel.ino

line 973:

` bool state=(bitRead(*((uint64_t *)d), settings.centerButton));`

and at line 854, add 2x more reading code:

```
  i=0x80;
  do
  {  
      if (!digitalReadFast(HC165_PIN_DATA1))
         d[4]|=i; 
      if (!digitalReadFast(HC165_PIN_DATA2))
         d[6]|=i; 

      digitalWriteFast(HC165_PIN_SCK, 1);
      digitalWriteFast(HC165_PIN_SCK, 0);  
      
  } while(i>>=1);
  i=0x80;
  do
  {   
      if (!digitalReadFast(HC165_PIN_DATA1))
         d[5]|=i; 
      if (!digitalReadFast(HC165_PIN_DATA2))
         d[7]|=i; 
      
      digitalWriteFast(HC165_PIN_SCK, 1);
      digitalWriteFast(HC165_PIN_SCK, 0);
               
  } while(i>>=1);
``` 
(as you can see, this is just copypasted and slightly modified code from just above. It could be done a little more efficiently, but need some testing)

I did not mention few more changes (changing types from uint32_t to uint64_t) which are required for using analog shifter and analog-pin-to-button features, I hope you won't need that.
Also, GUI will still show first 32 buttons - GUI communication code is not affected here, and GUI also has 32 buttons hardcoded in it.

It should compile, though I cannot say if it will work right. As I said, I cannot test in hardware now :(

**Reply by Pawan419** [reply]
Thank you so much sir.. i will test it..  
first i used hc165. But my soldering iron fried it and i 
find more ics in shop.  So I used CD4021. And it works fine..
But I faced one issue ..  when I gave same 5 volt and ground to encoder and BTS driver...
Encoder use to fail so I used separate supply for BTS driver. 
And CD4021 had same issue coz of it.. 8th button whenever pressed. It use to activate all 8 buttons of CD4021. But it was fixed when u gave separate VCC and gnd.

**Reply by Pawan419** [reply]
But can u use only 1 data pin for 4 CD4021 ic?? Or I need to add more Data3 and data4
  i=0x80;
  do
  {  
      if (!digitalReadFast(HC165_PIN_DATA1))
         d[4]|=i; 
      if (!digitalReadFast(HC165_PIN_DATA2))
         d[6]|=i; 

      digitalWriteFast(HC165_PIN_SCK, 1);
      digitalWriteFast(HC165_PIN_SCK, 0);  
      
  } while(i>>=1);
  i=0x80;
  do
  {   
      if (!digitalReadFast(HC165_PIN_DATA1))
         d[5]|=i; 
      if (!digitalReadFast(HC165_PIN_DATA2))
         d[7]|=i; 
      
      digitalWriteFast(HC165_PIN_SCK, 1);
      digitalWriteFast(HC165_PIN_SCK, 0);
               
  } while(i>>=1);

**Comment by vsulako** [comment]
CD4021Bs are handled in another place of code. 
So, instead of line 854, add this at line 898:

```
  i=0x80;
  do
  {  
      if (!digitalReadFast(CD4021_PIN_DATA1))
         d[4]|=i; 
      if (!digitalReadFast(CD4021_PIN_DATA2))
         d[6]|=i; 
      digitalWriteFast(CD4021_PIN_SCK, 1);
      digitalWriteFast(CD4021_PIN_SCK, 0);    
	  
  } while(i>>=1);
  i=0x80;
  do
  {   
      if (!digitalReadFast(CD4021_PIN_DATA1))
         d[5]|=i; 
      if (!digitalReadFast(CD4021_PIN_DATA2))
         d[7]|=i; 
      
      digitalWriteFast(CD4021_PIN_SCK, 1);
      digitalWriteFast(CD4021_PIN_SCK, 0);    
  } while(i>>=1); 
```

>>But can u use only 1 data pin for 4 CD4021 ic?? Or I need to add more Data3 and data4

Yes you can. Chained shift registers act just like one long register, no matter if you chain 2, 4, or dozens of them, you still will need same wires as in case of 1 ic.
You repeatedly read data from first IC, and each time all other in chain gradually shift all their data to first bit which is read - that's why they're called shift registers.

To form a chain, you connect serial in pin (pin 11) of previous CD4021 to output Q8 (pin 3) of next one.

As example, in wiring schematics for CD4021 (https://github.com/vsulako/AFFBWheel/blob/master/images/buttons_CD4021B.png) you can see 2 chains of 2 register IC's each. I don't remember now why I added pullups for serial in pins, obviously it was needed. 
Just continue to chain registers in same manner.
